WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,233 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using WalletWasabi.Exceptions;
+using WalletWasabi.Extensions;
+using WalletWasabi.Helpers;
+using WalletWasabi.Interfaces.EventSourcing;
+
+namespace WalletWasabi.EventSourcing
+{
+	/// <summary>
+	/// Thread safe without locks in memory event repository implementation
+	/// </summary>
+	public class InMemoryEventRepository : IEventRepository
+	{
+		protected static readonly IReadOnlyList<WrappedEvent> EmptyResult = Array.Empty<WrappedEvent>().ToList().AsReadOnly();
+		protected static readonly IReadOnlyList<string> EmptyIds = Array.Empty<string>().ToList().AsReadOnly();
+
+		private readonly ConcurrentDictionary<
+			string /*aggregateType*/,
+			ConcurrentDictionary<
+				string /*aggregateId*/,
+				(
+					// SequenceId of the last event of this aggregate
+					long TailSequenceId,
+
+					// locked flag for appending into EventsBatches of this aggregate
+					bool Locked,
```

Comment:
Perhaps an `enum` would be more readable than a bool in the compare-update below, because instead of `true` in the middle of a tuple, it could be `Locked` and `Unlocked`. The other positions in the tuple come equipped with variable names.

Reply:
good idea